youtube_dl/__init__.py

```diff
@@ -141,7 +147,22 @@ def _real_main(argv=None):
             parser.error('auto number start must be positive or 0')
     if opts.usetitle and opts.useid:
         parser.error('using title conflicts with using video ID')
-    if opts.username is not None and opts.password is None:
+    if opts.username is not None and opts.password is None and opts.password_from_keyring is True:
+        # extract domain names, check if all videos are from the same domain.
+        all_domains = set(map(lambda url: compat_urllib_parse.urlparse(url).netloc, all_urls))
+        if len(all_domains) > 1:
+            parser.error('You passed URLs from more than one domain - supplying credentials from command line is not supported in this case.')
+        domain_name = all_domains.pop()
+        password = Secret.password_lookup_sync(LIBSECRET_SCHEMA, {"user-name": opts.username, "domain-name": domain_name}, None)
```

Comment:
No. Passwords should be retrieved at the time of usage, see `_get_login_info`.

Reply:
This is troublesome, because AFAICS libsecret can only return a single value (password) - there's no way to retrieve other attributes of a keyring record. And `_get_login_info` needs to return both username and password.  
This ties in with your other comment https://github.com/rg3/youtube-dl/pull/14674#discussion_r148960973  
Basically, `--keyring` would be used as alternative to `--password`/interactive password. Would the following behaviour be acceptable? It seems to be working, could be refined further.  
1) Just `--username` = password prompt.
2) `--username` + `--keyring` = libsecret tries to look up password by special key (e.g. youtube) and username and sets `opts.password`. If password not found in keyring - throw error (or maybe fall back on password prompt to set `opts.password`).
3) `--username` + `--keyring` + `--password` = throw error  .

The only alternative I see atm is storing both username and password in a keyring record's secret with some separator, but that will lead to no end of parsing issues...

Reply:
Well, some just cut corners like [this](https://github.com/tommie-lie/rhythmbox-pandora/blob/45fe8e755d78c04ddeb348bf09a377c94ec8832c/rhythmbox-pandora/PandoraAccount.py#L27). A better way is to store a formatted string (JSON/XML/whatever). However, it indeed looks like it's not designed for storing multiple values. That brings to something alike you propose apart from you should not set `opts.password` yourself. If no password is found in keyring an error should be raised since no interactive actions are allowed in embeddable part of the code.

Reply:
All the choices so far are a bit hacky:  
1) combination of `--username` and `--keyring` is just bad - can't handle videos from multiple domains passed to `youtube-dl`
2) Strings with `\n` separator are hard (if not impossible) to store in a keyring through `secret-tool`. It would be fine if we were storing that string from within `youtube-dl`, but @dstftw is against it.
3) XML has >2 characters that need escaping ([source](https://stackoverflow.com/questions/730133/invalid-characters-in-xml))
4) JSON has 2 characters that need escaping - `"` and `\` ([source](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf)), we can instruct users to do that in README.md. JSON module is included in Python 2.6, same as `youtube-dl`'s lowest Python requirement.  

JSON looks like the most passable. What do you think of https://github.com/rg3/youtube-dl/pull/14674/commits/3be4a1db6b6e144390cb337888daba6c4b52c852?